src/memray/_memray/snapshot.cpp

```diff
@@ -192,6 +200,215 @@ TemporaryAllocationsAggregator::getSnapshotAllocations(bool merge_threads)
+}
+
+void
+HighWaterMarkAggregator::refreshUsageHistory(HighWaterMarkAggregator::UsageHistory& history)
+{
+    if (history.last_known_peak == d_peak_count) {
+        return;
+    }
+
+    // Any deltas since the last peak are part of the new one.
+    history.last_known_peak = d_peak_count;
+    history.allocations_contributed_to_last_known_peak += history.count_since_last_peak;
+    history.bytes_contributed_to_last_known_peak += history.bytes_since_last_peak;
+
+    history.count_since_last_peak = 0;
+    history.bytes_since_last_peak = 0;
+}
+
+void
+HighWaterMarkAggregator::recordPossiblePeak()
+{
+    if (d_current_heap_size < d_heap_size_at_last_peak) {
+        return;
+    }
+
+    // Found a new peak!
+    d_peak_count += 1;
+    d_heap_size_at_last_peak = d_current_heap_size;
+}
+
+void
+HighWaterMarkAggregator::recordUsageDelta(
+        const Allocation& allocation,
+        int64_t count_delta,
+        int64_t bytes_delta)
+{
+    if (bytes_delta < 0) {
+        // Update our stats if we're falling off a peak.
+        recordPossiblePeak();
+    }
+
+    d_current_heap_size += bytes_delta;
+
+    auto& history = getUsageHistory(allocation);
+    history.count_since_last_peak += count_delta;
+    history.bytes_since_last_peak += bytes_delta;
+}
+
+void
+HighWaterMarkAggregator::addAllocation(const Allocation& allocation)
+{
+    switch (hooks::allocatorKind(allocation.allocator)) {
+        case hooks::AllocatorKind::SIMPLE_ALLOCATOR: {
+            recordUsageDelta(allocation, 1, allocation.size);
+            d_ptr_to_allocation[allocation.address] = allocation;
+            break;
+        }
+        case hooks::AllocatorKind::SIMPLE_DEALLOCATOR: {
+            auto it = d_ptr_to_allocation.find(allocation.address);
+            if (it != d_ptr_to_allocation.end()) {
+                recordUsageDelta(it->second, -1, -static_cast<int64_t>(it->second.size));
+                d_ptr_to_allocation.erase(it);
+            }
+            break;
+        }
+        case hooks::AllocatorKind::RANGED_ALLOCATOR: {
+            d_mmap_intervals.addInterval(allocation.address, allocation.size, allocation);
+            recordUsageDelta(allocation, 1, allocation.size);
+            break;
+        }
+        case hooks::AllocatorKind::RANGED_DEALLOCATOR: {
+            auto removal_stats = d_mmap_intervals.removeInterval(allocation.address, allocation.size);
+            for (const auto& [interval, allocation] : removal_stats.freed_allocations) {
+                recordUsageDelta(allocation, -1, -static_cast<int64_t>(interval.size()));
+            }
+            for (const auto& [interval, allocation] : removal_stats.shrunk_allocations) {
+                recordUsageDelta(allocation, 0, -static_cast<int64_t>(interval.size()));
+            }
+            for (const auto& [interval, allocation] : removal_stats.split_allocations) {
+                recordUsageDelta(allocation, 1, -static_cast<int64_t>(interval.size()));
+            }
```

Comment:
I know, I just couldn't think of a better name ¯\\\_(ツ)\_/¯

OK. This code is a bit tricky to read, since the function parameter is called `allocation`, but might actually represent a deallocation, and if it does we need to look up the location(s) and size(s) of the allocation(s) being freed. I'm gonna handle this by renaming the function parameter to `const Allocation& allocation_or_deallocation`, and making every branch create a reference called `allocation` that always refers to the allocation being made or freed, and never refers to the deallocation record (which doesn't contain a location or size).